Feature/one to many

mstysk opened the pull request:
#47 

テストケースを作成しました。

既存の `Todo` に 複数の `Memo` を持たせるようにして、`OneToMany` の関係が出来たらいいと思いました。

どうでしょうか。

```mermaid

---
title: Test Entities
---
classDiagram
  Todo <-- Memo
  class Todo{
    string id
    string title
    Memo[] memos
  }
  class Memo{
    string id
    string body
  }

```

mstysk committed:
chore: tmp dir set ignore

mstysk committed:
chore: Memo sql created

mstysk committed:
feat: create testOneToMany

NaokiTsuchiya commented on tests/DbQueryModuleTest.php:
```diff
@@ -261,4 +264,20 @@ public function testFactory(string $queryInterface): void
+     */
+    public function testOneToMany(): void
+    {
+        $query = $this->injector->getInstance(TodoEntityInterface::class);
```
@mstysk
この場合、Memo をクエリするために実行したい SQL はどうなりますか？
SQL を追加、変更せずに Memo をクエリすることはできないと思うのですが。

NaokiTsuchiya replied:
issue の内容からは以下のようなSQLになるんでしょうか。

```sql
SELECT
  todo.id AS id, todo.title AS title, memo.id AS memo_id, memo.body AS memo_body
FROM todo
INNER JOIN memo on memo.todo_id = todo.id
WHERE todo.id = :id;
```

mstysk replied:
そうですね。上記のSQLみたいな形になるかもです。

もしくは、INNER JOIN ではなくて OUTER JOIN にしてもいいかもしれないですね。

memo が存在しない todo もあると思うので。

```sql
SELECT
  todo.id AS id, todo.title AS title, memo.id AS memo_id, memo.body AS memo_body
FROM todo
  LEFT OUTER JOIN memo on memo.todo_id = todo.id
WHERE todo.id = :id;
```

```sh
sqlite> SELECT * FROM todo;
1|run
2|sample
sqlite> SELECT * FROM memo;
1|run|1
sqlite> SELECT
   ...>   todo.id AS id, todo.title AS title, memo.id AS memo_id, memo.body AS memo_body
   ...> FROM todo
   ...> INNER JOIN memo on memo.todo_id = todo.id;
1|run|1|run
sqlite> SELECT
   ...>   todo.id AS id, todo.title AS title, memo.id AS memo_id, memo.body AS memo_body
   ...> FROM todo
   ...>   LEFT OUTER JOIN memo on memo.todo_id = todo.id;
1|run|1|run
2|sample||
sqlite>
```

mstysk replied:
こちらテストケースにsqlを追加しました。

koriym replied:
@mstysk 
そのクエリーを`PDO::FETCH_ASSOC`で行った時にこういう結果が返ってきますが。ここからMemoをどうやって作るかが課題でしょうか。

```
Array
(
    [0] => Array
        (
            [id] => 1
            [title] => run
            [memo_id] => 1
            [memo_body] => memo1
        )

    [1] => Array
        (
            [id] => 1
            [title] => run
            [memo_id] => 2
            [memo_body] => memo2
        )

)
```

```
Array
(
    [0] => Array
        (
            [id] => 1
            [title] => run
            [memo_id] => 1
            [memo_body] => memo1
        )

    [1] => Array
        (
            [id] => 1
            [title] => run
            [memo_id] => 2
            [memo_body] => memo2
        )

)
```

mstysk committed:
chore: todo join memo sql を作成

koriym commented:
@mstysk 今のSQLを実行するとこういう結果になります。

```
    [0] => Array
        (
            [id] => 1
            [title] => run
            [memo_id] => 1
            [memo_body] => memo1
        )

    [1] => Array
        (
            [id] => 1
            [title] => run
            [memo_id] => 2
            [memo_body] => memo2
        )
```

どういうルールでMemoを作るとかのルールが必要でしょうか

mstysk commented:
@koriym 

> どういうルールでMemoを作るとかのルールが必要でしょうか

多分そうだと思います。
以下のような `TodoEntity` が返却されると嬉しいです。

```
object(Ray\MediaQuery\Entity\Todo)#2 (3) {
  ["id"]=>
  int(1)
  ["title"]=>
  string(3) "run"
  ["memos"]=>
  array(2) {
    [0]=>
    object(Ray\MediaQuery\Entity\Memo)#4 (2) {
      ["id"]=>
      int(1)
      ["body"]=>
      string(5) "memo1"
    }
    [1]=>
    object(Ray\MediaQuery\Entity\Memo)#5 (2) {
      ["id"]=>
      int(2)
      ["body"]=>
      string(5) "memo2"
    }
  }
}
```

koriym committed:
Update SQL with GROUP_CONCAT

koriym committed:
Extract to CsvEnities

fixup! Extract to CsvEnities

fixup! Extract to CsvEnities

fixup! fixup! Extract to CsvEnities

koriym committed:
Change Memo factory support

koriym committed:
Todo to TodoConstruct

koriym commented:
以下の方針を立てて進めてみました。

* まずは実装を増やさないで既存の仕組みでできないか考えてみる
* JOINで得られたrowListの行数を、（joinするエンティティ=Memoの数だけ増えるのではなく）求めたい行数にするために、GROUP_CONCATを使う


```sql
SELECT
  todo.id AS id, 
  todo.title AS title,
  GROUP_CONCAT(memo.id),
  GROUP_CONCAT(memo.body)
FROM todo
LEFT OUTER JOIN memo
ON memo.todo_id = todo.id
GROUP BY todo.id;
```

```php
/** @return array<TodoMemo> */
public function getListWithMemo(): array;
```

```php
class TodoMemo
{
    /** @var array<Memo> */
    public array $memos;

    public function __construct(
        public string $id,
        public string $title,
        string|null $memoIds,
        string|null $memoBodies
    ){
        $this->memos = (new CsvEnities())(Memo::class, $memoIds, $memoBodies);
    }
}
````


TodoMemoにはSQLで得られたGROUP_CONCATによる連結されたCSVで渡されるので、それをエンティティオブジェクトにするCsvEnitiesクラスを利用し `array<Memo>`のリストをを作ります。

以下、自己分析です。

cons

* SQLにGROUP_CONCATが必要
* コンストラクタにボイラープレートコードが必要

pros

* エンティティ生成が何をしてるか明快
* コンストラクタでMemo以外のエンティティを操作したりなどのプログラム余地
* 実装コードに何の追加も変更もない

koriym committed:
Change factory to constructor to create csv entities

mstysk commented:
👍 LGTM

自分はこれでいいとおもいます！対応ありがとうございます！

koriym commented:
@NaokiTsuchiya どうでしょうか

koriym committed:
Switch to koriym/csv-entities

koriym commented:
Ray.MediaQueryとは完全に独立した機能なので、独立したパッケージにてそれを利用するようにしました。

https://github.com/koriym/Koriym.CsvEntities

koriym committed:
fixup! Switch to koriym/csv-entities

It's not minimum requirement.

mstysk commented:
ちょうど欲しいと思っていたので助かります。
